url2114d537a2_ForNeVeR_lanterna_tgz-pJ8-com_googlecode_lanterna_test_gui_DifferentKindsOfListBoxesTestJ8/src/com/googlecode/lanterna/terminal/text/UnixTerminal.java

```diff
@@ -0,0 +1,180 @@
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Copyright (C) 2010-2012 Martin
+ */
+
+package com.googlecode.lanterna.terminal.text;
+
+import com.googlecode.lanterna.LanternaException;
+import com.googlecode.lanterna.input.GnomeTerminalProfile;
+import com.googlecode.lanterna.input.PuttyProfile;
+import com.googlecode.lanterna.terminal.TerminalSize;
+import java.io.*;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.nio.charset.Charset;
+
+/**
+ * A common ANSI terminal extention with support for Unix resize signals and 
+ * the stty program to control cbreak and key echo
+ * @author Martin
+ */
+public class UnixTerminal extends ANSITerminal
+{
+    private final UnixTerminalSizeQuerier terminalSizeQuerier;
+            
+    public UnixTerminal(
+            InputStream terminalInput, 
+            OutputStream terminalOutput, 
+            Charset terminalCharset)
+    {
+        this(terminalInput, terminalOutput, terminalCharset, null);
+    }
+            
+    public UnixTerminal(
+            InputStream terminalInput, 
+            OutputStream terminalOutput, 
+            Charset terminalCharset,
+            UnixTerminalSizeQuerier customSizeQuerier)
+    {
+        super(terminalInput, terminalOutput, terminalCharset);
+        this.terminalSizeQuerier = customSizeQuerier;
+        addInputProfile(new GnomeTerminalProfile());
+        addInputProfile(new PuttyProfile());
+
+        //Make sure to set an initial size
+        onResized(80, 20);
+        try {
+            Class signalClass = Class.forName("sun.misc.Signal");
+            for(Method m: signalClass.getDeclaredMethods()) {
+                if("handle".equals(m.getName())) {
+                    Object handler = Proxy.newProxyInstance(getClass().getClassLoader(), new Class[] {Class.forName("sun.misc.SignalHandler")}, new InvocationHandler() {
+                        @Override
+                        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+                            if("handle".equals(method.getName())) {
+                                queryTerminalSize();
+                            }
+                            return null;
+                        }
+                    });
+                    m.invoke(null, signalClass.getConstructor(String.class).newInstance("WINCH"), handler);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)